### PR TITLE
Explicit null handling

### PR DESCRIPTION
**Pull request by gortiz**
This is another approach to implement column level nullability, as explained in https://github.com/apache/pinot/issues/10381.

# Introducing null handling modes

After discussing with @Jackie-Jiang, @walterddr and @npawar we decided to propose this third approach where each table specify how to handle nulls. There are two possible ways to handle null:
- Table mode, which is what we were using so far
- Column mode, which means that each column specifies its own nullability in the FieldSpec.

Multi-stage engine (aka V2) has also been modified to reject queries where `SET enableNullHandling = true` AND at least one of the tables involved in the query uses Table mode null handling.

## TODO
- [x] Modifications in `Schema`
- [x] Modifications in `TypeFactory`
- [x] Modifications in `NullValueIndexType`
- [x] Modifications in the ingestion side
- [x] Modifications in test engine
- [x] Write resource tests

## Proposed Syntax

Although the null handling mode semantics are clear, it is not so clear how to configure them. What is explained here is just a proposal open to discussion and can be changed during the PR review. 

The proposal is to add a new JSON property in the schema\*.

This new property can be configured as:
```js
{
  "schemaName": "blablabla",
  "dimensionFieldSpecs": [...],
  "enableColumnBasedNullHandling": true/false
}
```

The default value for `enableColumnBasedNullHandling` is false, which keeps compatibility with previous code. When it is changed to true, Pinot will ignore `TableConfig.IndexingConfig.nullHandlingEnabled` and columns will be nullable **if and only if** `FieldSpec.notNull` is false, which is also the default value.

\* It is yet to be decide whether it will be added to the TableConfig or the Schema. When this PR was open it was added to the Schema.

### CODE REVIEW

**Comment by walterddr** [comment]
CC @ankitsultana to also take a look

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java [hunk 1]
```diff
@@ -99,6 +100,7 @@ public abstract class FieldSpec implements Comparable<FieldSpec>, Serializable {
   protected String _name;
   protected DataType _dataType;
   protected boolean _isSingleValueField = true;
+  protected Boolean _nullable = null;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:103]: Per our discussion, should we add it as primitive boolean `_notNull` and default to false?

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 2]
```diff
@@ -57,13 +59,16 @@
...
+ *
+ * <p>The schema can be configured with different {@link Options}, including how to handle nulls</p>
  */
 @SuppressWarnings("unused")
 @JsonIgnoreProperties(ignoreUnknown = true)
 public final class Schema implements Serializable {
   private static final Logger LOGGER = LoggerFactory.getLogger(Schema.class);
 
   private String _schemaName;
+  private Options _options = new Options();
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:71]: Can we keep it simple by just adding a `boolean enableColumnBasedNullHandling` and default to false? This should be a short lived flag for migration, so I'd suggest keeping it simple

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 3]
```diff
@@ -520,14 +535,60 @@ public SchemaBuilder setSchemaName(String schemaName) {
...
+      consumer.accept(_schema._options);
+      return this;
+    }
+
+    public SchemaBuilder addMetricField(String name, DataType dataType) {
+      return addMetricField(name, dataType, ignore -> {
+      });
+    }
+
+    public SchemaBuilder addMetricField(String name, DataType dataType, Consumer<MetricFieldSpec> customizer) {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:548]: I don't see much value for adding this flexible customizer. `addField(FieldSpec fieldSpec)` can be handy though. This customizer is more or less creating a `FieldSpec`

**Review by Jackie-Jiang — COMMENTED** [review]
Can we keep the change small, especially the changes in `pinot-spi`? We should be very careful on any change added to `pinot-spi` because they cannot be easily removed

File: pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java [hunk 4]
```diff
@@ -45,20 +46,36 @@ public TypeFactory(RelDataTypeSystem typeSystem) {
...
-  private RelDataType toRelDataType(FieldSpec fieldSpec) {
+  private RelDataType toRelDataType(FieldSpec fieldSpec, Predicate<FieldSpec> isNullable) {
+    RelDataType type = createSqlType(getSqlTypeName(fieldSpec));
+    boolean isArray = !fieldSpec.isSingleValueField();
+    if (isArray) {
+      type = createArrayType(type, -1);
+    }
+    if (isNullable.test(fieldSpec)) {
+      type = createTypeWithNullability(type, true);
+    }
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java:69]: do we consider `isNullable` flag here refer to the array itself or this is refering to the element inside the array can be null? 
it is interesting b/c some DB consider array to be null <=> empty; but postgres do not (e.g. can have null array)
let's make sure we document it properly here.

File: pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java [hunk 5]
```diff
@@ -81,7 +83,7 @@ public void testQueryExplainPlansWithExceptions(String testCaseName, String quer
     }
   }
 
-  @DataProvider
+  @DataProvider(parallel = true)
```
Comment (walterddr) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java:86]: what's the reason behind this change? (btw. calcite is [not threadsafe](https://lists.apache.org/thread/wqjjsnyrwxsv9czctkh74v4r8h9w3w2r))

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java [hunk 6]
```diff
@@ -147,6 +147,13 @@ private static ServerPlanRequestContext build(OpChainExecutionContext executionC
     // 2. set pinot query options according to requestMetadataMap
     updateQueryOptions(pinotQuery, executionContext);
 
+    if (QueryOptionsUtils.isNullHandlingEnabled(pinotQuery.getQueryOptions())
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java:150]: i am not fully sure i understand the logic here. but I think it is OK to move this to a separate PR (not going to change the behavior if something is broken; and whatever works still works)

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:103]: We have run into problems with default true configs before, so it is best practice to use false as default to prevent bugs because people usually assume the default is false

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 3 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:548]: It is very valuable in order to keep the fluent dsl of the builder. There are several examples in this PR. Basically it is quite easier to write something like:

``` 
Schema schema = new SchemaBuilder()
   .addMetricField("name", type, field -> {
      //do your changes here
   })
   .build();
```

Than:

``` 
FieldSpec metric = new MetricField("name", type);
// do your changes here
Schema schema = new SchemaBuilder()
   .addField(field)
.build();
```

That is clear by the fact we have several builder methods like:
```
 public SchemaBuilder addSingleValueDimension(String dimensionName, DataType dataType);
 public SchemaBuilder addSingleValueDimension(String dimensionName, DataType dataType, Object defaultNullValue);
public SchemaBuilder addSingleValueDimension(String dimensionName, DataType dataType, int maxLength,
        Object defaultNullValue);
///etc
```

The current method does not scale while having a consumer is cleaner, shorter, scalable and safer. The later comes from the fact that having so many overloaded methods end up being problematic once we have a collision in types, while the consumer method is always type safe and explicit given we use named setters.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java (hunk 4 above)
Comment (gortiz) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java:69]: Most databases do not correctly support arrays (ie mysql 6.x). Postgres, on the other hand, has a sane type system (that even supports inheritance!). AFAIK when a column of type array is nullable it means the array itself is nullable. I don't know if each element in the array can be tuned nullable or not.

File: pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java (hunk 5 above)
Comment (gortiz) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java:86]: Calcite is not problematic here given we create a new instance on each query. It has to be thread safe in that case, otherwise two concurrent queries may affect each other.

This change reduce the time to test in my local computer by half. Given it is a free performance gain, I included it here and I encourage to use parallel whenever is possible.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java (hunk 6 above)
Comment (gortiz) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java:150]: The logic here is:
```
If a query is executed with V2 engine
  And is executed with null handling
  And any table consumed by the query is not using column level nullability
Then abort the execution with an error saying column null handling must be enable for all tables involved in a multi-stage query with null handling
```

I've been asked several times about queries returning strange results in V2 when using null handling. We don't support that but users don't know it. Therefore I strongly suggest to fail fast in order to be explicit about this limitation. This will also help users to understand that is important to migrate to column level nullability.

Comment (gortiz) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java:150]: But I'm ok removing this from the PR give it side affects several tests which are not actually testing nullability

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 2 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: I like the proposed approach, but I've simplified to what Jackie proposed in 11e597f4fe776c53ebaafbef3bed3c68446c0b96

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java (hunk 6 above)
Comment (gortiz) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java]: Removed in ac5c5e51c8ce8d418cc7cf35590e1a4e5a273a55

File: pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java (hunk 5 above)
Comment (walterddr) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java:89]: oh. i see. so this runs 2 individual processes? (calcite do share some static objects IIUC)

File: pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java (hunk 4 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java:69]: yeah agreed. as long as we document & best effort align with postgers it should be good :-D

File: pinot-integration-tests/src/test/resources/test_null_handling.schema [hunk 7]
```diff
@@ -1,4 +1,10 @@
 {
+  "options": {
+    "nullHandling": {
+      "mode": "column",
+      "default": true
```
Comment (walterddr) [inline pinot-integration-tests/src/test/resources/test_null_handling.schema:5]: - assume ***mode*** can either be `column` or `table`?
- so we plan to allow users to configure the ***default nullability*** on a table?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java [hunk 8]
```diff
@@ -67,7 +70,7 @@ public NullValueVectorCreator createIndexCreator(IndexCreationContext context, I
 
   @Override
   public IndexConfig getDefaultConfig() {
-    return IndexConfig.DISABLED;
+    return IndexConfig.ENABLED;
```
Comment (walterddr) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java:73]: just to confirm:
1. this means null index type is always enabled, but it could contain no columns?
2. this has nothing to do with the schema definition JSONs `option.nullHandling.mode/default` value yes?

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java [hunk 9]
```diff
@@ -300,6 +302,30 @@ public void setTransformFunction(@Nullable String transformFunction) {
     _transformFunction = transformFunction;
   }
 
+  /**
+   * Returns whether the column is nullable or not.
+   */
+  @JsonIgnore
+  public Boolean isNullable() {
```
Comment (walterddr) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:309]: primitive boolean

File: pinot-query-runtime/src/test/resources/queries/CountDistinct.json [hunk 10]
```diff
@@ -119,11 +119,13 @@
       {
         "comments": "table aren't actually partitioned by val thus all segments can produce duplicate results, thus [[8]]",
         "sql": "SELECT SEGMENT_PARTITIONED_DISTINCT_COUNT(val) FROM {tbl1}",
+        "ignored": true,
```
Comment (walterddr) [inline pinot-query-runtime/src/test/resources/queries/CountDistinct.json:122]: what's the reason behind this?

**Review by walterddr — COMMENTED** [review]
looks good to me mostly. please kindly see my comments. and some questions on the testing side.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java (hunk 8 above)
Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java:73]: > 1. this means null index type is always enabled, but it could contain no columns?

I don't think so. Columns contain indexes, not the other way arround

> 2. this has nothing to do with the schema definition JSONs option.nullHandling.mode/default value yes?

Nothing to do.

The reason to add this change is that before this change we always read the buffer, independently of the actual default config. Now we don't do that and therefore it is important to keep backward compatibility. Also, there are tons of tests that load segments without providing schema and table, which means that they use the default index config for all index types.

File: pinot-query-runtime/src/test/resources/queries/CountDistinct.json (hunk 10 above)
Comment (gortiz) [reply pinot-query-runtime/src/test/resources/queries/CountDistinct.json]: Not this one, but the other tests ignored fail just because I added more tables. It seem the test framework is not able to specify how to partition data and by adding tests with new tables we modify the default partition distribution.

You can try to remove the `ignored`  tests here and run the tests with and without my new tests and you will see that the results are different

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java (hunk 9 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:309]: fixed

**Comment by gortiz** [comment]
There are still some test failing. I wasn't able to understand why. I'll try to dedicate them more time tomorrow.

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 11]
```diff
@@ -520,14 +531,60 @@ public SchemaBuilder setSchemaName(String schemaName) {
...
+    }
+
     /**
      * Add single value dimensionFieldSpec
      */
     public SchemaBuilder addSingleValueDimension(String dimensionName, DataType dataType) {
       _schema.addField(new DimensionFieldSpec(dimensionName, dataType, true));
       return this;
     }
-
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:530]: (nit) revert

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java [hunk 12]
```diff
@@ -44,6 +44,7 @@
  * <p>- <code>IsSingleValueField</code>: single-value or multi-value field.
  * <p>- <code>DefaultNullValue</code>: when no value found for this field, use this value. Stored in string format.
  * <p>- <code>VirtualColumnProvider</code>: the virtual column provider to use for this field.
+ * <p>- <code>Nullable</code>: whether the column is nullable or not. Defaults to null
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:47]: Update this

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java [hunk 13]
```diff
@@ -99,6 +100,7 @@ public abstract class FieldSpec implements Comparable<FieldSpec>, Serializable {
   protected String _name;
   protected DataType _dataType;
   protected boolean _isSingleValueField = true;
+  protected boolean _nullable = true;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:103]: Since we use `notNull` as the key, can we also store `_notNull` internally and default to `false`? It is quite confusing to see it defined as `_nullable`, default to `true`, and has `isNullable()` and `setNullable()` json ignored

File: pinot-query-runtime/src/test/resources/queries/CountDistinct.json (hunk 10 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/CountDistinct.json]: ah i see. i recently fixed this in #11937 . PTAL

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java [hunk 14]
```diff
@@ -317,6 +343,7 @@ public ObjectNode toJsonObject() {
     }
     appendDefaultNullValue(jsonObject);
     appendTransformFunction(jsonObject);
+    jsonObject.put("notNull", !_nullable);
```
Comment (walterddr) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:346]: this would be super error-prone going forward IMO

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 11 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:530]: I already explained this is an improvement. I would prefer to do not remove it.

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java (hunk 13 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:103]: We can store that attribute. But adding too many negations makes things difficult to understand. It is easier to read
```
if (field.isNullable()) {
```

Than

```
if (!field.isNotNull()) {
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java (hunk 14 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java]: Fixed in 802e9c4f93caa0815e74d85035847c1a2153f20f

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java (hunk 13 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java]: Changed in 802e9c4f93caa0815e74d85035847c1a2153f20f

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java (hunk 12 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java]: Fixed in ccf1ef0f42aaaad7cdb7145813b88e360aae1ded

File: pinot-query-runtime/src/test/resources/queries/CountDistinct.json (hunk 10 above)
Comment (gortiz) [reply pinot-query-runtime/src/test/resources/queries/CountDistinct.json]: I've merged my branch with master and it didn't fix the issue. In fact now we have a third test here that fails. I'm going to remove the ignores so the issue is more evident.

**Comment by gortiz** [comment]
Marked as ready to review. There are 3 test failing, but they are the ones defined in CountDistinct.json that (AFAIK) are incorrectly failing

**Comment by gortiz** [comment]
In https://github.com/apache/pinot/actions/runs/6874613568/job/18696611890?pr=11960 you can see the test that failed. I'm going to comment them again in order to have a green light.

**Comment by walterddr** [comment]
> In https://github.com/apache/pinot/actions/runs/6874613568/job/18696611890?pr=11960 you can see the test that failed. I'm going to comment them again in order to have a green light.

the output value was changed in #11234 CC @Jackie-Jiang to take a look.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java [hunk 15]
```diff
@@ -73,6 +73,7 @@ public class ResourceBasedQueriesTest extends QueryRunnerTestBase {
   private static final String FILE_FILTER_PROPERTY = "pinot.fileFilter";
   private static final String IGNORE_FILTER_PROPERTY = "pinot.runIgnored";
   private static final int DEFAULT_NUM_PARTITIONS = 4;
+  private static final boolean REPORT_IGNORES = true;
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java:76]: can we separate this one from this PR?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java [hunk 16]
```diff
@@ -197,14 +199,17 @@ protected List<Object[]> queryH2(String sql)
     return result;
   }
 
-  protected void compareRowEquals(ResultTable resultTable, List<Object[]> expectedRows) {
-    compareRowEquals(resultTable, expectedRows, false);
+  protected void compareRowEquals(ResultTable resultTable, List<Object[]> expectedRows, String sql) {
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java:202]: not sure this change is necessary. in mvn tests the query will be part of the test name.
if we want to be more verbose here suggest take a stab into how test suite utilizes the parameterized test input

File: pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java (hunk 5 above)
Comment (walterddr) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java:89]: i wasn't sure this is fully verified. if possible let's keep this separate into another PR

**Review by walterddr — APPROVED** [review]
looks good to me mostly. please kindly address the follow ups. mostly on changes that can be separate into different PR

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java (hunk 15 above)
Comment (gortiz) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java:76]: Sure

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java (hunk 16 above)
Comment (gortiz) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java:202]: I've added that because in Intellij query is not added, which makes it difficult to find the query that fail. Also the description, if defined, should be part of the test name

Comment (gortiz) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java:202]: Sorry, I though you were complaining about the name of the tests. In this case... I definitively missed the query in the assertion.

File: pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java (hunk 5 above)
Comment (gortiz) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java:89]: I think it runs in quite more threads than just two. As said I noticed a large performance gain when running with this option. Neither I or the CI found any issue, but as requested I'll be conservative and remove it

Comment (gortiz) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/queries/ResourceBasedQueryPlansTest.java]: Done in e73aac404b600f2944324ae42833c974bf2dfe36

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 11 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:530]: I mean the unintended removal of empty line

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java (hunk 16 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java:202]: ah. i dont use intellij ui. it is printed in the test report from mvn test :-)

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 11 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: Fixed in 7716fe00bdb7f6c3bada07922deb39ec6b528e37

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java (hunk 16 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryRunnerTestBase.java:202]: but regardless, i think bottomline is can we agree to separate into a different PR (all the test improvements)

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java (hunk 15 above)
Comment (gortiz) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java]: I've rollbacked most of these changes in 3dd08862ed9e5c2355f029276675ffc2179be886

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java [hunk 17]
```diff
@@ -77,11 +80,23 @@ public String getPrettyName() {
...
-                : IndexConfig.DISABLED));
+    return (TableConfig tableConfig, Schema schema) -> {
+      Collection<FieldSpec> allFieldSpecs = schema.getAllFieldSpecs();
+      Map<String, IndexConfig> configMap = Maps.newHashMapWithExpectedSize(allFieldSpecs.size());
+
+      boolean enableColumnBasedNullHandling = schema.isEnableColumnBasedNullHandling();
+
+      for (FieldSpec fieldSpec : allFieldSpecs) {
+        IndexConfig indexConfig;
+        if (!enableColumnBasedNullHandling || fieldSpec.isNullable()) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java:91]: For my understanding, is this to keep the old behavior of always loading the null vector when the index exists?
In the existing code, we do check `tableConfig.getIndexingConfig().isNullHandlingEnabled()` but the check is dropped here. Is that because we do not check whether it is enabled in the existing code?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java [hunk 18]
```diff
@@ -550,6 +548,16 @@ private void createDefaultValueColumnV1Indices(String column)
...
 
+  private boolean isNullable(FieldSpec fieldSpec) {
+    if (_schema.isEnableColumnBasedNullHandling() && fieldSpec.isNullable()) {
+      return true;
+    } else {
+      return _indexLoadingConfig.getTableConfig() != null
+          && _indexLoadingConfig.getTableConfig().getIndexingConfig() != null
+          && _indexLoadingConfig.getTableConfig().getIndexingConfig().isNullHandlingEnabled();
+    }
+  }
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java:559]: This is incorrect. We want to ignore table based null handling when column based is enabled
```suggestion
    if (_schema.isEnableColumnBasedNullHandling()) {
      return fieldSpec.isNullable();
    } else {
      return _indexLoadingConfig.getTableConfig() != null
          && _indexLoadingConfig.getTableConfig().getIndexingConfig() != null
          && _indexLoadingConfig.getTableConfig().getIndexingConfig().isNullHandlingEnabled();
    }
  }
```

File: pinot-integration-tests/src/test/resources/test_null_handling.schema (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-integration-tests/src/test/resources/test_null_handling.schema:5]: Does this still apply? Should we set "enableColumnBasedNullHandling" instead?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java (hunk 17 above)
Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java:91]: You are right, although in my understanding nobody actually used this to decide whether the index is created/consumed or not. Even then, I'll change this code to be sure it is still backward compatible

Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexType.java]: Fixed in 9402a7c4a65a9d031f66a696031bc4154d217915. I think it is also easier to understand now.

File: pinot-integration-tests/src/test/resources/test_null_handling.schema (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-integration-tests/src/test/resources/test_null_handling.schema:5]: ^^
Also, how does it work right now? Do we have test that rely on column based null handling to be enabled?

File: pinot-query-planner/src/test/java/org/apache/pinot/query/testutils/MockRoutingManagerFactory.java [hunk 19]
```diff
@@ -107,6 +113,16 @@ public TableCache buildTableCache() {
       String schemaName = invocationOnMock.getArgument(0);
       return _schemaMap.get(schemaName);
     });
+    when(mock.getTableConfig(anyString())).thenAnswer(invocationOnMock -> {
+      String tableNameWithType = invocationOnMock.getArgument(0);
+      String tableName = TableNameBuilder.extractRawTableName(tableNameWithType);
+      return new TableConfigBuilder(TableType.OFFLINE)
+          .setTableName(tableName)
+//          .setSchemaName(_schemaMap.get(tableName).getSchemaName())
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/testutils/MockRoutingManagerFactory.java:121]: (nit) Remove the commented line

File: pinot-query-runtime/src/test/resources/queries/CountDistinct.json (hunk 10 above)
Comment (Jackie-Jiang) [reply pinot-query-runtime/src/test/resources/queries/CountDistinct.json:122]: Shall we try removing the ignored flag and see if they still fail? If so, we might want to debug them because this can cause regression

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-query-runtime/src/test/resources/queries/CountDistinct.json (hunk 10 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/CountDistinct.json:122]: i tried it and it didnt' work. let's address separately

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexTypeTest.java [hunk 20]
```diff
@@ -50,6 +50,7 @@ protected void assertEquals(IndexConfig expected) {
 
     @Test(dataProvider = "provideCases", dataProviderClass = NullValueIndexTypeTest.class)
     public void isEnabledWhenNullable(boolean columnNullHandling, boolean fieldNullable, IndexConfig expected) {
+      _tableConfig.getIndexingConfig().setNullHandlingEnabled(true);
```
Comment (walterddr) [inline pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/nullvalue/NullValueIndexTypeTest.java:53]: this is needed b/c the logic now rely on table config. (backward incompat?)
